The plugin "gatsby-source-datocms" deleted a node of a type owned by   another plugin.

Hi,

After upgrading Gatsby to version 2.22.10 and using this plugin on version 2.2.4, whenever I try to build my app without removing the `.cache/datocms-assets` folder, build fails with the following stack trace:

```
  Error: The plugin "gatsby-source-datocms" deleted a node of a type owned by   another plugin.
            The node type "DatoCmsField" is owned by "undefined".
            The node object passed to "deleteNode":
            {
      "id": "DatoCmsField-677586",
      "parent": null,
      "children": [],
      "internal": {
          "type": "DatoCmsField",
          "contentDigest": "d8805d2fdf57b367fb9b78ab4d884bd0",
          "counter": 246,
          "owner": "gatsby-source-datocms"
      },
      "label": "Title",
      "fieldType": "string",
      "apiKey": "title",
      "localized": true,
      "validators": {
          "required": {}
      },
      "position": 1,
      "appeareance": {
          "addons": [],
          "editor": "single_line",
          "parameters": {
              "heading": false
          }
      },
      "defaultValue": {
          "fr": null,
          "en": null
      },
      "originalId": "677586"
  }
            The plugin deleting the node:
            {
      "resolve": "/Users/jimmy/projects/ector-front/node_modules/gatsby-sourc  e-datocms",
      "id": "408ce464-e94c-5d22-a673-28525d31a655",
      "name": "gatsby-source-datocms",
      "version": "2.2.4",
      "pluginOptions": {
          "plugins": [],
          "apiToken": "-------",
          "apiUrl": "https://site-api.datocms.com",
          "disableLiveReload": true,
          "previewMode": true
      },
      "nodeAPIs": [
          "createSchemaCustomization",
          "sourceNodes"
      ],
      "browserAPIs": [],
      "ssrAPIs": [],
      "pluginFilepath": "/Users/jimmy/projects/ector-front/node_modules/gatsb  y-source-datocms"
  }
  
  - public.js:485 actions.deleteNode
    [ector-front]/[gatsby]/dist/redux/actions/public.js:485:70
  
  - redux.js:483 
    [ector-front]/[redux]/lib/redux.js:483:35
  
  - api-runner-node.js:90 Object.doubleBoundActionCreators.<computed> [as del    eteNode]
    [ector-front]/[gatsby]/dist/utils/api-runner-node.js:90:20
  
  - destroyEntityNode.js:44 
    [ector-front]/[gatsby-source-datocms]/hooks/sourceNodes/destroyEntityNode    .js:44:23
  
  - Array.map
  
  - destroyEntityNode.js:40 module.exports
    [ector-front]/[gatsby-source-datocms]/hooks/sourceNodes/destroyEntityNode    .js:40:19
  
  - index.js:158 
    [ector-front]/[gatsby-source-datocms]/hooks/sourceNodes/index.js:158:15
  
  - Array.forEach
  
  - EntitiesRepo.js:130 
    [ector-front]/[datocms-client]/lib/local/EntitiesRepo.js:130:25
  
  - Array.forEach
  
  - EntitiesRepo.js:129 EntitiesRepo.destroyEntities
    [ector-front]/[datocms-client]/lib/local/EntitiesRepo.js:129:29
  
  - EntitiesRepo.js:90 
    [ector-front]/[datocms-client]/lib/local/EntitiesRepo.js:90:18
  
  - Array.forEach
  
  - EntitiesRepo.js:88 EntitiesRepo.destroyAllEntities
    [ector-front]/[datocms-client]/lib/local/EntitiesRepo.js:88:56
  
  - Loader.js:93 
    [ector-front]/[datocms-client]/lib/local/Loader.js:93:29
  

```

Okay so God knows why but if I run
$ yarn add gatsby-transformer-remark

I can then run yarn develop successfully, only once, and it will post a minor error:

/Users/blain/blaininc.github.io/src/components/layout.js
  93:19  warning  A control must be associated with a text label  jsx-a11y/control-has-associated-label

✖ 1 problem (0 errors, 1 warning)

BUT the localhost 8000 will stand up successfully

I cannot run yarn develop a second time, without again running
$ yarn add gatsby-transformer-remark

Once I run yarn add gatsby-transformer-remark again
I can then run
$ yarn build

which then runs a successful build operation with no errors posted.

Bizarre. Why would this work? Why would it stop working unless I reinstall that component again? What is happening there?

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "gatsby-source-datocms",
-  "version": "2.2.3",
+  "version": "2.2.4",
   "description": "Gatsby source plugin for building websites using DatoCMS as data source",
   "main": "index.js",
   "scripts": {

--- a/package.json
+++ b/package.json
@@ -23,7 +23,7 @@
   },
   "dependencies": {
     "core-js": "^3.2.1",
-    "datocms-client": ">=3.0.24",
+    "datocms-client": ">=3.0.31",
     "fs-extra": ">=8.1.0",
     "humps": ">=2.0.1",
     "imgix-url-params": "11.2.0",

--- a/package.json
+++ b/package.json
@@ -23,7 +23,7 @@
   },
   "dependencies": {
     "core-js": "^3.2.1",
-    "datocms-client": ">=3.0.31",
+    "datocms-client": ">=3.0.32",
     "fs-extra": ">=8.1.0",
     "humps": ">=2.0.1",
     "imgix-url-params": "11.2.0",

--- a/src/hooks/createSchemaCustomization/index.js
+++ b/src/hooks/createSchemaCustomization/index.js
@@ -1,8 +1,9 @@
 const fs = require('fs-extra');
-const finalizeNodesCreation = require('../sourceNodes/finalizeNodesCreation');
+const createNodeFromEntity = require('../sourceNodes/createNodeFromEntity');
+const destroyEntityNode = require('../sourceNodes/destroyEntityNode');
 const createTypes = require('../sourceNodes/createTypes');
 
-const { getClient, getLoader } = require('../../utils');
+const { getLoader } = require('../../utils');
 
 module.exports = async (
   { actions, getNode, getNodesByType, reporter, parentSpan, schema, store },
@@ -38,8 +39,18 @@ module.exports = async (
 
   activity.start();
 
+  const removeUpserListener = loader.entitiesRepo.addUpsertListener(entity => {
+    createNodeFromEntity(entity, context);
+  });
+
+  const removeDestroyListener = loader.entitiesRepo.addDestroyListener(entity => {
+    destroyEntityNode(entity, context);
+  });
+
   await loader.loadSchema();
-  finalizeNodesCreation(context);
+
+  removeUpserListener();
+  removeDestroyListener();
 
   activity.end();
 

--- a/src/hooks/sourceNodes/index.js
+++ b/src/hooks/sourceNodes/index.js
@@ -106,6 +106,9 @@ module.exports = async (
     return;
   }
 
+  let activity = reporter.activityTimer(`loading DatoCMS content`, { parentSpan });
+  activity.start();
+
   loader.entitiesRepo.addUpsertListener(entity => {
     createNodeFromEntity(entity, context);
   });
@@ -114,11 +117,6 @@ module.exports = async (
     destroyEntityNode(entity, context);
   });
 
-  let activity;
-
-  activity = reporter.activityTimer(`loading DatoCMS content`, { parentSpan });
-
-  activity.start();
   await loader.load();
   finalizeNodesCreation(context);
 

